File: 2019/aula-14/josephus.cpp
#include <cstdio>

const int maxn = 35100;
bool is_prime[maxn];
int prime[maxn];
int prime_count;

struct node_t {
  int value;
  node_t *next;
};
node_t* make_list(int n) {
  node_t* start = new node_t;
  node_t* ptr = start;
  start->value = n;
  for(int i=1; i<n; ++i) {
    ptr->next = new node_t;
    ptr = ptr->next;
    ptr->value = i;
  }
  ptr->next = start;
  return start;
}
void remove_next(node_t* ptr) {
  node_t* tmp = ptr->next;
  ptr->next = tmp->next;
  delete tmp;
}
void generate_primes() {
  is_prime[0]=is_prime[1]=false;
  prime_count=0;
  for(int i=2; i<maxn; ++i)
    is_prime[i]=true;
  for(int i=2; i<maxn; ++i) {
    if (is_prime[i]) {
      prime[prime_count++]=i;
      for(int j=i*i; j<maxn; j+=i)
        is_prime[j]=false;
    }
  }
}
bool alive[maxn];

int solve_old(int n) {
  for(int i=0; i<n; ++i) {
    alive[i]=true;
  }
  int current=0;
  for(int i=0; i<n-1; ++i) {
    int skip = prime[i]%(n-i);
    if (skip==0)
      skip+=n-i;
    while(skip) {
      current=(current+1)%n;
      if (alive[current])
        --skip;
    }
    alive[current]=false;
  }
  for(int i=1; i<n; ++i) {
    if (alive[i])
      return i;
  }
  return n;
}
int solve(int n) {
  node_t* list = make_list(n);
  for(int i=0; i<n-1; ++i) {
    int skip = (prime[i]-1)%(n-i);
    for(int j=0; j<skip; ++j)
      list = list->next;
    remove_next(list);
  }
  int ans = list->value;
  delete list;
  return ans;
}
int main() {
  int tmp;
  generate_primes();
  while(true) {
    scanf(" %d", &tmp);
    if (tmp==0)
      break;
    printf("%d\n", solve(tmp));
  }
  return 0;
}

File: 2019/aula-14/josephus.py
class Node:
    def __init__(self, value):
        self.next_ptr = None
        self.value = value

def make_list(n):
    start_ptr = Node(n)
    ptr = start_ptr
    for i in range(1, n):
        ptr.next_ptr = Node(i)
        ptr = ptr.next_ptr
    ptr.next_ptr = start_ptr
    return start_ptr

def remove_next(list_ptr):
    list_ptr.next_ptr = list_ptr.next_ptr.next_ptr

def solve(n, prime):
    list_ptr = make_list(n)
    for i in range(n-1):
        skip = (prime[i]-1)%(n-i)
        for j in range(skip):
            list_ptr = list_ptr.next_ptr
        remove_next(list_ptr)
    return list_ptr.value

def generate_primes():
    limit = 35_100
    number_list = list(range(limit))
    number_list[0] = number_list[1] = False
    primes = []
    for i in number_list:
        if i:
            primes.append(i)
            j = i*i
            while j<limit:
                number_list[j] = False
                j = j+i
    return primes

def main():
    prime = generate_primes()
    print(len(prime))
    while True:
        x = int(input())
        if x==0:
            break
        print(solve(x, prime))

main()
